src/cargo/ops/cargo_install.rs

```diff
@@ -182,16 +197,48 @@ fn install_one(
             }
         }
         src.update()?;
-        select_pkg(src, krate, vers, config, false, &mut |path| {
+        select_pkg(&mut src, krate, None, config, false, &mut |path| {
             path.read_packages()
         })?
     } else {
+        let mut source = map.load(source_id, &HashSet::new())?;
+        let vers = if let Some(vers) = vers {
+            Some(parse_semver_req(vers)?)
+        } else if source.source_id().is_registry() {
+            Some(VersionReq::any())
```

Comment:
I'm a bit confused by this check.  Why is it needed if the code below immediately checks for `is_exact`?

Reply:
There is something really subtle here, so I think it might be worth changing this.  From a naive look, `select_pkg` with a version of None should use the "any" requirement implicitly as `parse_no_deprecated` [does this](https://github.com/rust-lang/cargo/blob/805462e971db208d338691cced1ad396b01450e2/src/cargo/core/dependency.rs#L190).  However, due to the round-trip of string->VersionReq->string->VersionReq, the "any" requirement gets changed to `"*"` which is *different* from `VersionReq::any`, as `"*"` means "do not select pre-release versions" (see https://github.com/steveklabnik/semver/issues/161).  This is behavior Cargo relies on.

I'd like this to be a bit more explicit.  I'm thinking, change it back so that `select_pkg` takes `Option<&str>`, and structure it more like it used to be.  That way, there is an explicit step (where the old comment `Avoid pre-release versions` was) that can very clearly call out this behavior. That way there isn't this distant assumption about what `VersionReq::any` means.  `installed_exact_package` can throw away whatever work it does to determine if it is an exact version, I don't think it helps to pass in a `VersionReq` to select_pkg.